### PR TITLE
Add OutFile property in WebResponseObject

### LINKED ISSUE #21082: `Invoke-WebRequest -PassThru`, when combined with `-OutFile`, should reflect the download file path.
### Summary of the new feature / enhancement

Thanks to a contribution by @CarloToso, since v7.4.0 it has been possible to pass a _directory_  rather than a file path to `-OutFile`, in which the web cmdlets use the (ultimate) target URL's _last segment_ as the file name.

Given that, when redirections are involved,  you won't know what that last segment is, being able to _programmatically_ discover the resulting full download file path is important.

This will also apply to the planned support for using a  _server-suggested_ output file name via a `Content-Disposition` response header field - see PR https://github.com/PowerShell/PowerShell/pull/19385

(Using `-Verbose` is a way to discover the file name _interactively_; while it's technically possible to programmatically parse the verbose output, that would be both cumbersome and brittle).

Thus - at least when `-OutFile` is used - the `[Microsoft.PowerShell.Commands.WebResponseObject]` instance returned via `-PassThru` should have an `.OutFile` property or similar that reflects the output file's full path.

Related:

 * #21081

### Proposed technical implementation details (optional)

I don't know whether it's worth extending `[Microsoft.PowerShell.Commands.WebResponseObject]` or not.

If not, an ETS `NoteProperty` member could be used.

### PR DESCRIPTION
**Pull request by jshigetomi**
# PR Summary

Fixes: #21082 

Added OutFile as a property in WebResponseObject. OutFile will save whenever writing to pipe line.
Refactored one line to reduce redundancy.

## PR Context

PR extends WebObjectResponse with a property that stores the installed file's path. It makes sense to have this as a built in property since we have access to it.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): 
- **User-facing changes**
  - [x] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
  - [x] N/A or can only be tested interactively
  - **OR**
  - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: 
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: 
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: 
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs [hunk 1]
```diff
@@ -70,6 +70,11 @@ public class WebResponseObject
         /// </summary>
         public string StatusDescription => WebResponseHelper.GetStatusDescription(BaseResponse);
 
+        /// <summary>
+        /// Gets the out file path
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs:74]:
```suggestion
        /// Gets or sets the output file path.
```

**Review by iSazonov — COMMENTED** [review]
Please add tests for the scenario in `WebCmdlets.Tests.ps1`.

**Comment by iSazonov** [comment]
@mklement0 Please review.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs [hunk 2]
```diff
@@ -70,6 +70,11 @@ public class WebResponseObject
         /// </summary>
         public string StatusDescription => WebResponseHelper.GetStatusDescription(BaseResponse);
 
+        /// <summary>
+        /// Gets or sets the output file path.
+        /// </summary>
+        public string? OutFile { get; set; }
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs:76]: should this be an internal set?

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs (hunk 2 above)
Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs:76]: I agree (`public string? OutFile { get; internal set; }`)

@SteveL-MSFT, @iSazonov, just to confirm: Is it generally fine to extend public classes with new members?

Other than that, LGTM - thanks for taking this on, @jshigetomi.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs:76]: @mklement0 Removing public element is a breaking change, adding is ok.

**Comment by iSazonov** [comment]
@jshigetomi Please resolve merge conflicts.

**Review by iSazonov — APPROVED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs (hunk 2 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs:76]: @jshigetomi Please address the request.

Comment (jshigetomi) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebResponseObject.Common.cs]: @iSazonov Thank you for bringing this to my attention. I misread this chain and thought there were no actionable items.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs [hunk 3]
```diff
@@ -38,6 +38,8 @@ internal override void ProcessResponse(HttpResponseMessage response)
             ArgumentNullException.ThrowIfNull(response);
             TimeSpan perReadTimeout = ConvertTimeoutSecondsToTimeSpan(OperationTimeoutSeconds);
             Stream responseStream = StreamHelper.GetResponseStream(response, _cancelToken.Token);
+            string outFilePath = WebResponseHelper.GetOutFilePath(response, _qualifiedOutFile);
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs:41]: `_qualifiedOutFile` is null when `ShouldSaveToOutFile == false`, so this line of code needs to be in the `if (ShouldSaveToOutFile ) { ... }` block, cannot be put here.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]
I guess the `Invoke-ResetMethod -OutFile -Passthru` needs to be updated accordingly to have the similar behavior, doesn't it?

https://github.com/PowerShell/PowerShell/blob/8016daaddab76f59aeafc04b48bec5a12f1b9fe2/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs#L143

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs (hunk 3 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs:41]: GetOutFilePath() returns null if _qualifiedOutFile is null. It doesn't any extra work. I see no point in complicating the code with an additional conditional block.

**Comment by iSazonov** [comment]
> I guess the `Invoke-ResetMethod -OutFile -Passthru` needs to be updated accordingly to have the similar behavior, doesn't it?

I didn't find `WebResponseObject` there. It is seem used only in `Invoke-WebRequest`.

---

But I found `WebResponseObject` and `BasicHtmlWebResponseObject` in PowerShellCore_format_ps1xml.cs. I believe it would be useful to add new `Outfile` property in the formatting.

**Review by SteveL-MSFT — APPROVED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs (hunk 3 above)
Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs:41]: Good point. Resolved.

**Comment by daxian-dbw** [comment]
> I guess the Invoke-ResetMethod -OutFile -Passthru needs to be updated accordingly to have the similar behavior, doesn't it?

The `Invoke-RestMethod` does support `-OutFile` and `-Passthru` to be specified together, so the file information should be kept in the returned object just like `Invoke-WebRequest`, doesn't it?

**Comment by mklement0** [comment]
@daxian-dbw: `Invoke-RestMethod -PassThru -OutFile` is currently _broken_:

* #15409

If it did / once it does work as intended, what would be passed through is (possibly parsed into a `[pscustomobject]` graph / an XML DOM) _data_, not a `WebResponseObject` instance.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs [hunk 4]
```diff
@@ -61,8 +64,6 @@ internal override void ProcessResponse(HttpResponseMessage response)
 
             if (ShouldSaveToOutFile)
             {
-                string outFilePath = WebResponseHelper.GetOutFilePath(response, _qualifiedOutFile);
-
                 WriteVerbose(string.Create(System.Globalization.CultureInfo.InvariantCulture, $"File Name: {Path.GetFileName(outFilePath)}"));
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/InvokeWebRequestCommand.CoreClr.cs:67]:
```suggestion
                WriteVerbose($"File Name: {Path.GetFileName(outFilePath)}");
```

**Review by daxian-dbw — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@mklement0 Thanks for pointing to that issue. I didn't know it's currently broken. I re-opened it.

**Pull request merged**